evm_arithmetization/src/generation/mod.rs

```diff
@@ -201,23 +223,85 @@ fn apply_metadata_and_tries_memops<F: RichField + Extendable<D>, const D: usize>
             .collect::<Vec<_>>(),
     );
 
+    // Write initial registers.
+    let registers_before = [
+        registers_before.program_counter,
+        registers_before.is_kernel,
+        registers_before.stack_len,
+        registers_before.stack_top,
+        registers_before.context,
+        registers_before.gas_used,
+    ];
+    ops.extend((0..registers_before.len()).map(|i| {
+        mem_write_log(
+            channel,
+            MemoryAddress::new(0, Segment::RegistersStates, i),
+            state,
+            registers_before[i],
+        )
+    }));
+
+    let length = registers_before.len();
+
+    // Write final registers.
+    let registers_after = [
+        registers_after.program_counter,
+        registers_after.is_kernel,
+        registers_after.stack_len,
+        registers_after.stack_top,
+        registers_after.context,
+        registers_after.gas_used,
+    ];
+    ops.extend((0..registers_before.len()).map(|i| {
+        mem_write_log(
+            channel,
+            MemoryAddress::new(0, Segment::RegistersStates, length + i),
+            state,
+            registers_after[i],
+        )
+    }));
+
     state.memory.apply_ops(&ops);
     state.traces.memory_ops.extend(ops);
 }
 
-pub fn generate_traces<F: RichField + Extendable<D>, const D: usize>(
+type TablesWithPVsAndFinalMem<F> = (
+    [Vec<PolynomialValues<F>>; NUM_TABLES],
+    PublicValues,
+    Vec<Vec<F>>,
+);
+pub(crate) fn generate_traces<F: RichField + Extendable<D>, const D: usize>(
```

Comment:
One should be extremely cautious with this kind of breaking changes. The method was until a few days ago used externally by the test runner. There should always be motivation as to why restricting such method visibility.

Reply:
I made a pass and removed all the changes in visibility (whenever possible).